<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" />
    <title>Online Prediciton of Human Whole Genomic Linkages</title>
    <link href="./CSS/style1.css" rel="stylesheet" type="text/css" />
    <!-- <script language="javascript" type="text/javascript" src="./JS/moreinfo.js"></script> -->
  </head> 
  
  <body>
    <h3><p>Online Prediciton of Human Whole Genomic Linkages</p></h3>
    
    
    <!-- Input candidate data -->
    <form target="_blank" method="post" action="/cgi-bin/phyloCGI/linkage_predict.py" enctype="multipart/form-data">
      <fieldset>
	<legend><h2>Candidate Gene List</h2></legend>
	<p>Choose batch file: <input type="file" name="candidateList"/></p>
      </fieldset>
      
      <br />
      <center>
	<p><input type="submit" value="Submit" /></p>
      </center>
    </form>

    <!-- example -->
    <hr />
    <h3>Example</h3>
    <table>
      <tr>
	<td>Click to see example:</td>
	<td><a href="./Exampledata/phylopred_atpSub/index.html" target="_blank"><input type="button" value="Example"/></a></td>
	<td>&nbsp;</td>
      </tr>
      <tr>
	<td>Click to download example input files (F1Fo ATP synthase):</td>
	<td><a target="_blank" href="./Exampledata/phylopred_atpSub/atpSub.txt">"txt" format</a></td>
	<td><a target="_blank" href="./Exampledata/phylopred_atpSub/atpSub.csv">"csv" format</a></td>
      </tr>
    </table>

    <!-- help information -->
    <hr />
    <h3>User's guide</h3>
    <h4>Input</h4>
    <p>Only the "txt" or the "csv" format is allowed for the input files. </p>
    <ol>
      <li>The number of input genes is supposed to be at least two. There are serveral reasons. The most important one is our algorithm is based on the whole genome, and the prediction results are more reliable when taking into account a cluster of genes like genes from a complex or a biological pathway. Our team is now working on an advanced algorithms based on the statistical distribution of linkages for each gene. Several clues indicate the new algorithm have a better performace in reducing the false positive rate, which haunts all the existing method for protein linkage prediction. </li>
      <li>1st column contains the input genes. The "GeneID" is used as the unique gene name for each gene, for example "hsa:513" denotes the human gene "ATP5D" (ATP synthase, H+ transporting, mitochondrial F1 complex, delta subunit (EC:3.6.1.14)). Other formats of gene names may be supported in the furture. </li>
      <li>2nd column contains the colours for each input genes. Colour modes like the hex triplet and colour names are both supported. For example, the "red" and "#ff0000" are legal input colours. If you do not want to show the colour for a gene, an easy way is to set the colour as "white" or "#FFFFFF". </li>
    </ol>
    
    <h4>Ouput</h4>
    <p>We provide two ways to illustrate the prediction results. A webpage merged the all the figures and tables will be returned at first. At the bottom of the webpage, a linkage will guild you to download all the results into your local computer.</p>
    <ol>
      <li>The webpage contains visualization of phylogenetic profiles, correlation matrix, and list of prediction linkages.</li>
      <li>The downloaded folder includes published quality of figures method above in both "pdf" and "jpg" formats. The prediction linkages is stored in a "csv" file that can be easily used for other analysis and validation experiments. Moreover, the same webage is stored in a "html" file used for a overview of the results.</li>
    </ol>
    
    
    
    <!-- contact information -->
    <hr />
    <p>If you have any questions or suggestion, please do not hesitate to contact us. </p>
    <h4>Yulong Niu</h4>
    <p>email: <a href="mailto:yulong.niu@aol.com">yulong.niu (AT) aol.com</a></p>
    
  </body>


  <!-- google anlaytics -->
  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-53073828-3', 'auto');
  ga('send', 'pageview');

  </script>
  
</html>
